### ISSUE TITLE
bug with editing events after timezone change

### ISSUE THREAD

**Issue by fritzmg** [contao/core#8101]
### Short description

If you change the timezone from e.g. Europe/Vienna (UTC+1) to Europe/London (UTC) and then edit an event, the event date will be one __day__ (and one hour, which is correct) in the past of the original date.

### Reproduction

* set the timezone to _Europe/Vienna_
* create a new event, set the start date to `2015-11-12` and the start time (and end time) to `08:00`, save and close
* go to the system settings and set the timezone to _Europe/London_
* go to the calendar - the overview will display the event's date correctly as `[2015-11-12 07:00]`
* edit the event - now the start date will be 2015-11-__11__ for some reason (start time will be 07:00 which is correct)

This does not happen in the other direction, i.e., if you

* set the timezone to _Europe/London_
* create a new event with the start date 2015-11-12 and start/end time 08:00
* set the timezone to _Europe/Vienna_
* edit the event again

the start date will still be 2015-11-12 and the start/end time will be 09:00.

**Comment by fritzmg** [comment]
The cause is, that Contao saves the dates always with the time _00:00:00_. i.e. a start date for 2015-11-12 will be `2015-11-12 00:00:00 UTC+1` (as a UNIX timestamp). This would be `2015-11-11 23:00:00 UTC` of course. It's weird though, that the event list in the backend still shows `2015-11-12` after the change.

Maybe the start and end _date_ should always be saved as `YYYY-MM-DD 12:00:00`? (and then always re-convert the date to `YYYY-MM-DD 00:00:00` and `YYYY-MM-DD 00:00:00 +1day` when checking for the actual start and end timestamp).

**Comment by fritzmg** [comment]
btw. in the event list of the Backend, the correct time is shown, because `startTime` is used to display the date. `startTime` consists of the whole timestamp, i.e. date + time, thus a minus one hour change of the timezone only affects the time (in this case)

I reckon when not using a time, the backend list would also show the "wrong" date.

**Comment by leofeyer** [comment]
@contao/developers /cc

**Comment by aschempp** [comment]
I just stumbled upon a new information regarding timestamps. The PHP `DateTime` class threads a timestamp as **always** UTC.

http://php.net/manual/de/datetime.construct.php, first comment
> There's a reason for ignoring the time zone when you pass a timestamp to __construct.  That is, UNIX timestamps are by definition based on UTC.  @1234567890 represents the same date/time regardless of time zone.  So there's no need for a time zone at all.

So apparently we're actually wrong to use *local* timestamps. But I have no idea how to fix all existing data…

**Comment by fritzmg** [comment]
> But I have no idea how to fix all existing data…

Well it's not just about fixing the existing data. How should Contao save the actual `date`? Currently Contao uses the timestamp of the given date in the given timezone, which is correct - however, the time part of the date is saved as `00:00:00`, which leads to the problem described in this thread. This will always lead to problems. The only timezone safe way I can think of is to save the `date` as an ISO 8601 or RFC 2822 string and the `time` simply as a a `H:i:m` string (which would then be relative to the `date`'s timezone).

**Comment by aschempp** [comment]
Technically correct would be to store it in a `DATE` field in the database, or use an UTC timestamp.

**Comment by fritzmg** [comment]
> UTC timestamp.

Wouldn't that still lead to problems? If this "UTC" timestamp is used with the regular `date()` function of PHP, it might again be the wrong day. For instance, if your event is on the 2016-04-18, the timestamp that Contao would then save is `1460937600000`. If your timezone setting is something negative to UTC, this would yield `2016-04-17`.

**Comment by aschempp** [comment]
That's correct, we should not use `date()` but `DateTime`. Which again is a problem…

**Comment by leofeyer** [comment]
We have discussed this issue again in Mumble on July 21st and besides that it will be a lot of work to migrate everything to using `DateTime`, it will also be a huge BC break. Therefore the earliest possibility to change it is Contao 5.
